espnet2/train/spk_trainer.py

```diff
@@ -0,0 +1,185 @@
+
+            batch = to_device(batch, "cuda" if ngpu > 0 else "cpu")
+            if no_forward_run:
+                continue
+
+            org_shape = (batch["speech"].size(0), batch["speech"].size(1))
+            batch["speech"] = batch["speech"].flatten(0, 1)
+            batch["speech2"] = batch["speech2"].flatten(0, 1)
+
+            speech_embds = model(
+                speech=batch["speech"], spk_labels=None, extract_embd=True
+            )
+            speech2_embds = model(
+                speech=batch["speech2"], spk_labels=None, extract_embd=True
+            )
+
+            speech_embds = F.normalize(speech_embds, p=2, dim=1)
+            speech2_embds = F.normalize(speech2_embds, p=2, dim=1)
+
+            speech_embds = speech_embds.view(org_shape[0], org_shape[1], -1)
+            speech2_embds = speech2_embds.view(org_shape[0], org_shape[1], -1)
+
+            for i in range(speech_embds.size(0)):
+                score = torch.cdist(speech_embds[i], speech2_embds[i])
+                score = -1.0 * torch.mean(score)
+                scores.append(score.view(1))  # 0-dim to 1-dim tensor for cat
+            labels.append(batch["spk_labels"])
+
+        else:
+            if distributed:
+                iterator_stop.fill_(1)
+                torch.distributed.all_reduce(iterator_stop, ReduceOp.SUM)
+
+        scores = torch.cat(scores).type(torch.float32)
+        labels = torch.cat(labels).type(torch.int32).flatten()
+        if distributed:
+            # get the number of trials assigned on each GPU
+            length = to_device(
+                torch.tensor([labels.size(0)], dtype=torch.int32), "cuda"
+            )
+            lengths_all = [
+                to_device(torch.zeros(1, dtype=torch.int32), "cuda")
+                for _ in range(torch.distributed.get_world_size())
+            ]
+            torch.distributed.all_gather(lengths_all, length)
+
+            scores_all = [
+                to_device(torch.zeros(i, dtype=torch.float32), "cuda")
+                for i in lengths_all
+            ]
+            torch.distributed.all_gather(scores_all, scores)
+            scores = torch.cat(scores_all)
+
+            labels_all = [
+                to_device(torch.zeros(i, dtype=torch.int32), "cuda")
+                for i in lengths_all
+            ]
+            torch.distributed.all_gather(labels_all, labels)
+            labels = torch.cat(labels_all)
+            rank = torch.distributed.get_rank()
+            torch.distributed.barrier()
+        scores = scores.detach().cpu().numpy()
+        labels = labels.detach().cpu().numpy()
+
+        # exception for collect_stats.
+        if len(scores) == 1:
+            reporter.register(stats=dict(eer=1.0, mindcf=1.0))
+            return
+
+        results = tuneThresholdfromScore(scores, labels, [1, 0.1])
+        eer = results[1]
+        fnrs, fprs, thresholds = ComputeErrorRates(scores, labels)
+        p_trg, c_miss, c_fa = 0.05, 1, 1
```

Comment:
Can you add some comments to explain the numbers (`[1, 0.1]` and `0.05, 1, 1`) here?

Reply:
I will add in part 4. Hope you don't mind!

Reply:
OK. Just to comment here for convenience.

Reply:
Done.